clippy_lints/src/eta_reduction.rs

```diff
@@ -65,6 +65,9 @@ fn check_closure(cx: &LateContext<'_, '_>, expr: &Expr) {
             if !(is_adjusted(cx, ex) || args.iter().any(|arg| is_adjusted(cx, arg)));
 
             let fn_ty = cx.tables.expr_ty(caller);
+
+            if let ty::FnDef(_, _) = fn_ty.sty;
```

Comment:
This should include `FnPtr` and `Closure`. 

Use `matches!` if it works here, otherwise you can move this check below to the `then` block as a `match`